### ISSUE TITLE
Bug with year_ranges in combination of different regions.

### ISSUE THREAD

**Issue by knut2** [holidays/holidays#174]
Preamble:
This is a follow up of #173 and you can find my changes in https://github.com/knut2/holidays/tree/reformationday2017DE

Background:
In some countries in Germany the "Reformationstag" is a holiday.
In 2017 is the 500 jubilee of reformation and this holiday is valid for whole Germany.

So I defined in de.yaml:

```
  10:
  - name: Reformationstag
    regions: [de_bb, de_mv, de_sn, de_st, de_th]
    mday: 31
  - name: Reformationstag
    regions: [de_bw]
    mday: 31    
    type: informal
  - name: Reformationstag
    regions: [de]
    mday: 31    
    year_ranges:
    - limited: [2017]
```

The generator.rb generates:
```
      10 => [
            {:mday => 31, :name => "Reformationstag", :regions => [:de_bb, :de_mv, :de_sn, :de_st, :de_th, :de]},
            {:mday => 31, :type => :informal, :name => "Reformationstag", :regions => [:de_bw]}],
```

The year_range is ignored in this combination, :de is added as an additional region.

After a small change in generator.rb, line 78 (check also differences in :year_ranges) I get:

```
      10 => [
            {:mday => 31, :name => "Reformationstag", :regions => [:de_bb, :de_mv, :de_sn, :de_st, :de_th]},
            {:mday => 31, :type => :informal, :name => "Reformationstag", :regions => [:de_bw]},
            {:mday => 31,  :year_ranges => [{:limited => [2017]}],:name => "Reformationstag", :regions => [:de]}],
```

But if I run my test
```
assert_equal 'Reformationstag', Holidays.on(Date.civil(2017,10,31), :de)[0][:name]
assert_equal [], Holidays.on(Date.civil(2016,10,31), :de), "Reformationstag is not a holiday in 2016 in whole Germany"
assert_equal [], Holidays.on(Date.civil(2018,10,31), :de), "Reformationstag is not a holiday in 2018 in whole Germany"

```

I get a failure 
```
Failure: test_de(DeDefinitionTests)
test_defs_de.rb:40:in `test_de'
     37: end
     38: #500 years reformation in 2017
     39: assert_equal 'Reformationstag', Holidays.on(Date.civil(2017,10,31), :de)[0][:name]
  => 40: assert_equal [], Holidays.on(Date.civil(2016,10,31), :de), "Reformationstag is not a holiday in 2016 in whole Germany"
     41: assert_equal [], Holidays.on(Date.civil(2018,10,31), :de), "Reformationstag is not a holiday in 2018 in whole Germany"
     42: 
     43: [:de_bw, :de_by, :de_nw, :de_rp, :de_sl, :de_].each do |r|
Reformationstag is not a holiday in 2016 in whole Germany
<[]> expected but was
<[{:date=>#<Date: 2016-10-31 ((2457693j,0s,0n),+0s,2299161j)>,
  :name=>"Reformationstag",
  :regions=>[:de_bb, :de_mv, :de_sn, :de_st, :de_th, :de]}]>
```

Again, the year_range for :de is ignored.

**Comment by knut2** [comment]
Sometimes it helps to explain the problem: 
I have to do the change in the generator also in `Holidays::Definition::Repository::HolidaysByMonth#definition_exists?`

Now my test works correct.
I updated my branch.

**Comment by ttwo32** [comment]
@knut2 Sorry for failure with different regions.
That's a newish feature , so i am glad year_range is used in other region. 
I appreciate your refactoring.

**Comment by ptrimble** [comment]
Thank you @knut2! I appreciate your PR and the extra details you provide. I will be looking at this shortly. 

I am in the process of finishing up a rather large refactoring for https://github.com/holidays/holidays/issues/144. My plan is to handle all of the outstanding PRs (yours included) and then rebase against those changes on my branch. I will handle merge conflicts on my branch so others won't have to feel the pain of my refactoring.

It'll take me a few days to work through the existing PRs. I'll be in touch!

**Comment by ptrimble** [comment]
This has been addressed and will be released either with a `3.3.0` release alongside other definition updates or with the major version bump to `4.0.0`. Either way it will be released to the public in the next two weeks. Closing! Thanks @knut2!
